server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStoreSettings.java

```diff
@@ -0,0 +1,75 @@
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.regex.Regex;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.core.UpdateForV10;
+import org.elasticsearch.features.NodeFeature;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+import static org.elasticsearch.core.Predicates.never;
+
+/**
+ * Holder for the data stream global settings relating to the data stream failure store. This defines, validates, and monitors the settings.
+ */
+public class DataStreamFailureStoreSettings {
+
+    @UpdateForV10(owner = UpdateForV10.Owner.DATA_MANAGEMENT) // Remove this when we no longer need BWC with versions without this feature
+    static final NodeFeature NODE_FEATURE = new NodeFeature("data_streams.failure_store.cluster_setting");
+
+    private static final Logger logger = LogManager.getLogger(DataStreamFailureStoreSettings.class);
+
+    public static final Setting<List<String>> DATA_STREAM_FAILURE_STORED_ENABLED_SETTING = Setting.stringListSetting(
+        "data_streams.failure_store.enabled",
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    private Predicate<String> failureStoreEnabledByName;
+
+    private DataStreamFailureStoreSettings() {
+        failureStoreEnabledByName = never();
+    }
+
+    /**
+     * Creates an instance and initialises the cluster settings listeners.
+     *
+     * @param clusterSettings The cluster settings to initialize the instance from and to watch for updates to
+     */
+    public static DataStreamFailureStoreSettings create(ClusterSettings clusterSettings) {
+        DataStreamFailureStoreSettings dataStreamFailureStoreSettings = new DataStreamFailureStoreSettings();
+        clusterSettings.initializeAndWatch(
+            DATA_STREAM_FAILURE_STORED_ENABLED_SETTING,
+            dataStreamFailureStoreSettings::setEnabledByNamePatterns
+        );
+        return dataStreamFailureStoreSettings;
+    }
+
+    /**
+     * Returns whether the settings indicate that the failure store should be enabled by the cluster settings for the given name.
+     *
+     * @param name The data stream name
+     */
+    public boolean failureStoreEnabledForDataStreamName(String name) {
+        assert DataStream.isFailureStoreFeatureFlagEnabled() : "Testing whether failure store is enabled should be behind by feature flag";
+        return failureStoreEnabledByName.test(name);
+    }
+
+    private void setEnabledByNamePatterns(List<String> patterns) {
```

Comment:
Ah! Such a good point that I completely missed. Apologies. Actually, we should only include this setting in the `ClusterSettings` if the feature flag is enabled. This will have as an effect that this setting won't exist in the system and if someone defines it will throw an error no matter what.

You can see commit https://github.com/elastic/elasticsearch/commit/01ed7de99ff22f92f08db11b666be88794f33829#diff-8e9e5bbba5094b994150b78efa335859b3435167c688e5e21d079dc92c198e6d, for reference.

Reply:
Good call. I've pushed a commit which only registers the setting in `ClusterSettings` if the flag is enabled, and also changes `DataStreamFailureStoreSettings.create` to only `initializeAndWatch` the setting on the same condition as well (otherwise that call would fail).